File: 0020-valid-parentheses/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        # Approach: Stack -- T.C: O(n), S.C: O(n)

        n = len(s)

        # Optimization: Exit early
        if n % 2 == 1:
            return False

        stack = []
        for c in s:
            if c == "(":
                stack.append(")")
            elif c == "{":
                stack.append("}")
            elif c == "[":
                stack.append("]")
            else:
                if len(stack) == 0 or stack[-1] != c:
                    return False
                else:
                    stack.pop()
        return not stack

File: 0020-valid-parentheses/20.c
#include <string.h>
bool isValid(char *s) {
    /* Approach: Stack -- T.C: O(n), S.C: O(n) */

    int n   = strlen(s);
    int top = -1;

    char *stack = (char *)malloc((n + 1) * sizeof(char));

    for (int i = 0; i < n; i++) {
        char c = s[i];
        switch (c) {
        case '(':
            stack[++top] = ')';
            break;
        case '{':
            stack[++top] = '}';
            break;
        case '[':
            stack[++top] = ']';
            break;
        default:
            if (top == -1 || stack[top] != c)
                return false;
            else
                top--;
        }
    }

    free(stack);
    return top == -1;
}
